### PR TITLE
MapShed Manual Entry: Land Cover Modal

### PR DESCRIPTION
**Pull request by rajadain**
## Overview

Adds a Land Cover Modal to edit the current land cover distribution. You can change the value of any existing land cover in hectares, but the total area must remain the same, so other values must be changed accordingly to compensate. The changes cannot be saved until the total matches.

When the changes are saved, they are added as "under the hood" modifications that override values in the base GMS, just as other manual entries do.

The Land Cover modal opens from a separate button than other Settings, akin to TR-55, in an attempt to harmonize the two model types.

Connects #2970

### Demo

screen shot 2018-09-18 at 4 34 52 pm

screen shot 2018-09-18 at 4 34 58 pm

screen shot 2018-09-18 at 4 35 08 pm

screen shot 2018-09-18 at 4 35 59 pm

screen shot 2018-09-18 at 4 38 40 pm

## Testing Instructions

* Check out this branch and `bundle`
* Open a MapShed project. Make a new scenario.
* Open the Land Cover modal. Ensure the default values are there for each land cover type.
* Change the value of a field. Ensure you see a message indicating by how far your are off the total. Ensure you are unable to save the modal.
* Change another value to make the total match. Ensure you are now able to save the modal.
* Save the modal. Ensure calculations are re-run.
* Open the modal again. Ensure your changed values are preserved and populated.
* Refresh the page to reload the scenario. Open the modal again. Ensure your changed values are preserved and populated.
* Export Current Conditions and your scenario as GMS files and diff them. Ensure that the land cover areas have changed.

### CODE REVIEW

**Comment by jfrankl** [comment]
Visually, this looks good to me! 👍

**Comment by kellyi** [comment]
Code looks good to me. Trying out the functionality now.

**Comment by kellyi** [comment]
This looks really cool!

screen shot 2018-09-19 at 1 11 48 pm

Some quick feedback: it seems really hard to get the values to be exact right now. I even tweaked them to make them apparently exact per the validation instruction and still had the weird `(0 more than required...)` error.

Is there a way we can make the data entry a little more fuzzy and/or have an option to give all the remaining area to a selected land cover type?

**Comment by kellyi** [comment]
FWIW I can apparently cure that 0 more than expected error by just reducing one of the decimals by one one-thousandth of a percent:

screen shot 2018-09-19 at 1 18 19 pm

-> 

screen shot 2018-09-19 at 1 18 26 pm

Seems like we should at least make the inputs do some rounding down when necessary?

**Comment by kellyi** [comment]
One other quirky-ish thing is that when the validation message is applied, it isn't re-calculated until the active input loses focus, so it's possible to update the input to an acceptable value but not know that it's acceptable until after blurring the text box.

Is there a simple way to adjust this?

**Comment by kellyi** [comment]
Aside from the validation quirks this is working well!

**Comment by rajadain** [comment]
I pushed two commits: 3733432 should take care of the input responsiveness (see commit message for more details). 8d56141 moves the floating point logic out of the template into JS which is a little more reasonable. I think the above two issues are fixed, but they could use a second look.

**Comment by kellyi** [comment]
Cool, taking another look at this!

File: src/mmw/js/src/core/utils.js [hunk 1]
```diff
@@ -562,6 +562,11 @@ var utils = {
         return geom;
     },
 
+    // Round a number x to n decimal places
+    round: function(x, n) {
+        return Math.round(parseFloat(x) * Math.pow(10, n)) / Math.pow(10, n);
```
[the remark this answers is not in the document]
Comment (rajadain) [reply src/mmw/js/src/core/utils.js]: Good call.

[the remark this answers is not in the document]
Comment (rajadain) [reply src/mmw/js/src/core/utils.js]: Lodash 3.6.0 doesn't have it 😞 https://github.com/lodash/lodash/blob/3.6.0/lodash.js

**Comment by rajadain** [comment]
Thanks for taking a look! I'm going to squash the fixups and merge.

**Pull request merged**